### ISSUE TITLE
azuread_application_api_access OwnedBy permissions not working

### ISSUE THREAD

**Issue by ryan-royals** [hashicorp/terraform-provider-azuread#1460]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritise this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritise the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureAD Provider) Version

Terraform v1.9.3
+ provider registry.terraform.io/hashicorp/azuread v2.53.1

### Affected Resource(s)

* `azuread_application_api_access`

### Terraform Configuration Files

```hcl
data "azuread_application_published_app_ids" "well_known" {}

data "azuread_service_principal" "msgraph" {
  client_id = data.azuread_application_published_app_ids.well_known.result["MicrosoftGraph"]
}

resource "azuread_application_registration" "app1" {
  display_name = "azuread_application_registration_test"
}

resource "azuread_application_api_access" "app1" {
  application_id = azuread_application_registration.app1.id

  api_client_id = data.azuread_application_published_app_ids.well_known.result["MicrosoftGraph"]
  role_ids = [
    data.azuread_service_principal.msgraph.app_role_ids["Group.Read.All"],
  ]
}

```

### Apply Output

``` Text
module.app_reg.azuread_application_api_access.app1: Creating...
╷
│ Error: creating Application API Access (Application ID: "***", API Client ID: "00000003-0000-0000-c000-000000000000"): ApplicationsClient.BaseClient.Patch(): unexpected status 403 with OData error: Authorization_RequestDenied: Insufficient privileges to complete the operation.
│ 
│   with module.app_reg.azuread_application_api_access.app1,
│   on library/azure/application_registration/main.tf line 33, in resource "azuread_application_api_access" "app1":
│   33: resource "azuread_application_api_access" "app1" {
│ 
│ creating Application API Access (Application ID:
│ "***", API Client ID:
│ "00000003-0000-0000-c000-000000000000"):
│ ApplicationsClient.BaseClient.Patch(): unexpected status 403 with OData
│ error: Authorization_RequestDenied: Insufficient privileges to complete the
│ operation.
╵
```
### Expected Behavior

According to documentation, the OwnedBy permission should be enough to complete this operation.

### Actual Behavior

OwnedBy permissions is not allowing the deploy of `api_access`  when using Terraform, but is sufficient when using the Portal or AzCLI.

### Steps to Reproduce

1. `terraform apply` using a identity that is a `Application Developer` so that it becomes OwnedBy on deploy.

**Comment by nbaju1** [comment]
Is the identity also an owner of the application?

Ref from the docs: 
"When using the `Application.ReadWrite.OwnedBy` application role, the principal being used to run Terraform must be an owner of the application."

**Comment by ryan-royals** [comment]
> Is the identity also an owner of the application?
> 
> Ref from the docs: "When using the `Application.ReadWrite.OwnedBy` application role, the principal being used to run Terraform must be an owner of the application."

Yeah. The creation of the Application is being done by Terraform, and is allocated as the Owner on the Application.

**Comment by ryan-royals** [comment]
Thanks, that makes sense.
Seems like a area that could be improved though, considering that App Dev role works when using Portal or CLI.

I'd take a crack at it, but I don't have any programming skills so I wouldn't know where to start / if it is actually something that can be fixed with a PR.

**Comment by nbaju1** [comment]
It's probably a limitation in the Graph API itself, which is what the provider uses, so don't think there is really anything to do except use the suggested role.

**Comment by phillips756** [comment]
@ryan-royals @nbaju1 
Have either of you tried with only graph api permissions delegated for app registration and removing directory roles from the SPN? 

We are currently working with a vendor to deploy an app in our tenant that creates app registrations, SPs, and assigns them subscription based resources and roles etc. somewhat standard stuff you'd use Terraform for as far as I can tell. We believe that Application administrator should not be given to a third party app to manage all applications/SPs in a tenant when it only needs to manage a handful that it is is directly creating - that's exactly the intention of the app developer role + creator/owner workflows. We are having problems with the same issue mentioned here and are trying to determine if the issue is MS side or Terraform side. If @nbaju1 interpretation is correct of the language just being faulty, then removing directory roles but relying on graph only permission grants would potentially solve the issue. If it doesn't... then I believe this a pretty big security hole / problem and determining if the issue is the commands Terraform is ultimately passing to the API or MS Graph not respecting proper scopes for the endpoints.

**Comment by ryan-royals** [comment]
> @ryan-royals @nbaju1 Have either of you tried with only graph api permissions delegated for app registration and removing directory roles from the SPN?
> 
> We are currently working with a vendor to deploy an app in our tenant that creates app registrations, SPs, and assigns them subscription based resources and roles etc. somewhat standard stuff you'd use Terraform for as far as I can tell. We believe that Application administrator should not be given to a third party app to manage all applications/SPs in a tenant when it only needs to manage a handful that it is is directly creating - that's exactly the intention of the app developer role + creator/owner workflows. We are having problems with the same issue mentioned here and are trying to determine if the issue is MS side or Terraform side. If @nbaju1 interpretation is correct of the language just being faulty, then removing directory roles but relying on graph only permission grants would potentially solve the issue. If it doesn't... then I believe this a pretty big security hole / problem and determining if the issue is the commands Terraform is ultimately passing to the API or MS Graph not respecting proper scopes for the endpoints.

Since we are using Managed Identities in our environment, we can not (easily) use purely Graph API permissions. I have only tested with Directory roles.

**Comment by nbaju1** [comment]
> @ryan-royals @nbaju1 Have either of you tried with only graph api permissions delegated for app registration and removing directory roles from the SPN?
> 
> We are currently working with a vendor to deploy an app in our tenant that creates app registrations, SPs, and assigns them subscription based resources and roles etc. somewhat standard stuff you'd use Terraform for as far as I can tell. We believe that Application administrator should not be given to a third party app to manage all applications/SPs in a tenant when it only needs to manage a handful that it is is directly creating - that's exactly the intention of the app developer role + creator/owner workflows. We are having problems with the same issue mentioned here and are trying to determine if the issue is MS side or Terraform side. If @nbaju1 interpretation is correct of the language just being faulty, then removing directory roles but relying on graph only permission grants would potentially solve the issue. If it doesn't... then I believe this a pretty big security hole / problem and determining if the issue is the commands Terraform is ultimately passing to the API or MS Graph not respecting proper scopes for the endpoints.

We are using Graph API permissions only, no issues here.
